### DISCUSSION TITLE
Decoding CopyLock for Freescape game releases for Amiga/AtariST

### DISCUSSION THREAD

**Discussion by neuromancer** [keirf/disk-utilities#211] [Q&A] [answered]
Hi, I'm a ScummVM developer implementing the [freescape engine](https://wiki.scummvm.org/index.php?title=Freescape). Some of these games have releases that included several layers of copy protection, including copylock for some Amiga/AtariST releases of Dark Side, Total Eclipse and Castle Master. We want to offer instructions for owners of these games to dump their disks and extract the executables. ScummVM will take care of the rest of the protections, since we already reimplemented them. I recently found that `disk-utilities` allow to read/convert the copylock track from disk images, and you even include a `copylock` utility to emulate/disassembly the protection, however, it seems it does not save the decrypted executable. I'm wondering how much effort will be to do it, and if there is any way to help.

**Comment by keirf** [comment]
Are we talking about a Copylock-protected Amiga exe which gets decrypted at load time? It should be pretty easy. I've not looked at such a Copylock before but it may be as simple as remembering where segments were loaded in this loop https://github.com/keirf/disk-utilities/blob/master/m68k/copylock.c#L117 and then saving out those segments at the end of the Copylock emulation.

Tell me IPF(s) or whatever to take a look at, and I can see what I can do.

**Comment by neuromancer** [comment]
Hi,

> Are we talking about a Copylock-protected Amiga exe which gets decrypted at load time?

Yes.

> Tell me IPF(s) or whatever to take a look at, and I can see what I can do.

There is a copylocked ipf in archive.org [here](https://ia802208.us.archive.org/view_archive.php?archive=/4/items/commodore-amiga-games-sps/Dark%20Side%20v1.0%20%281989%29%28Micro%20Status%29%5B1099%5D.zip&file=DarkSide.ipf). The executable file that is copylocked is called 0.drk (you can check it in the header). It will decrypt itself using copylock and then load the other file 1.drk and decrypt it using some code that we already reimplemented, so we only need to get a clean version 0.drk.

Thanks in advance!

**Comment by keirf** [comment, marked as answer]
Here is the linked Dark Side disk, converted to ADF, and with the cracked 0.drk file inserted. If what you want is a process by which users can create their own cracked 0.drk file, then that is more involved.

[darkside_cracked.zip](https://github.com/user-attachments/files/16051451/darkside_cracked.zip)

**Reply by neuromancer** [reply]
Thanks! I will be taking a look to the cracked version. In the meantime, how hard is the process to dump it?

**Reply by neuromancer** [reply]
I can confirm that the cracked 0.drk file works without any changes in the current code :tada:

**Reply by keirf** [reply]
You mean, to generate the clean exe? Pretty easy. See the following folder: https://github.com/keirf/disk-utilities/tree/master/amiga/cracks/dark_side

Basically, Copylock overwrites itself after decrypting and applying relocations to the original executable code hunk. The copy happens always (it seems) in low memory, so we can breakpoint there and dump regs/memory to a file. Then there is a little Python script to scrape the relocation table and the decrypted code from the state file, un-apply the relocations, and emit a new exe file with a normal RELOC32 hunk.

**Reply by neuromancer** [reply]
This is exactly what we need (well, actually, we don't need a working binary, just one that has the data for decrypting the other file, which is saved as some kind of table). In any case, this is very useful, we can already start modifying our documentation for Dark Side in order to show users how to decrypt their files. Do you think this should be applicable for other Dark Side release with CopyLock (there are at least two in AtariST with this protection) or for other freescape games (e.g. Total Eclipse).

**Reply by keirf** [reply]
Other Amiga titles, yes. Other Atari ST titles sadly no. Because the tools would need modifying to emulate Atari ST hardware (for Copylock emulation) and generate a new ST executable. Obviously doable but outside my own knowledge and inclination.

But other Amiga titles: point me at them and I'll do them, and add them to the disk-utilities repo. It would then be easy enough for a Linux-literate tech-savvy user to regenerate the clean Amiga executables.

**Reply by keirf** [reply]
Regarding the steps for decryption, feel free to make any suggestions for making it friendlier or easier to document. But as you can see it's basically run "make" so pretty simple. Just the file naming maybe to tweak.

**Reply by keirf** [reply]
I checked the IPFs I have for Amiga versions of Driller and Total Eclipse: These do not have disk protection.

Castle Master claims to have Copylock but actually is a different Rob Northen protection hidden in AmigaDOS track 1. However, it's probably another encrypted executable, albeit the file (intro) seems to be naked and is raw-loaded by a preceding binary (cm).

**Reply by neuromancer** [reply]
Thanks a lot for your support!, I will check the scripts and perhaps suggest some little changes here and there in a small PR. Btw, I checked the other Amiga disk images and there is one more copylocked game: [Castle Master / The Crypt from the Virtual Worlds bundle](https://archive.org/download/commodore-amiga-games-sps/Castle%20Master%20%26%20Castle%20Master%20II%20-%20The%20Crypt%20%281990%29%28Incentive%29%5B0612%5D.zip).

**Reply by keirf** [reply]
Appears slightly trickier because the protection is embedded rather than at the start of an exe. Do you know what you actually need from Castle Master / The Crypt? The whdload crack could perhaps be applied. It doesn't look very dependent on whdload infra.

**Reply by neuromancer** [reply]
Coming back to this, the specific files needed for Castle Master / The Crypt are `cmstr.com` and `crypt.com`. They are supposed to be Amiga executables. Additionally, there is Space Station Oblivion, which was the alternative name for Driller in the US. The [ipf](https://archive.org/download/commodore-amiga-games-sps/Space%20Station%20Oblivion%20%281988%29%28Epyx%29%28US%29%5B1428%5D.zip) seems to be protected somehow, but I'm not sure if it is CopyLock or something else. Can you take a look? (if it is necessary to open another discussion for that, let me know and I will do it).

**Reply by keirf** [reply]
I don't think `cmstr.com` and `crypt.com` are straightforward executables. Actually I suspect they are stripped and encrypted. Again, you just need data tables out of these, actually?

Space Station Oblivion has looks like a straightforward long track. I doubt it's Rob Northen. Probably something simpler. It may not even involve encryption.

Someone pointed me at the following tool for Atari ST Copylocks and other protections: https://github.com/orionfuzion/dec0de
It may help!

**Reply by neuromancer** [reply]
Regarding  `cmstr.com` and `crypt.com`: I need the full binaries, since the data is actually mixed with the code. However, the code is actually not needed. If it is there, that's fine. But there is no need to have a functional executable. If you check the decrypted binaries, you should be able to see the full list of strings used in the games (as well as other things). For example:
```
***************
 PRESS ANY KEY 
  OVERPOWERED  
 YOU COLLAPSE  
    CRUSHED    
  FATAL FALL   
 YOU DID IT !  
```

Regarding:

> Space Station Oblivion has looks like a straightforward long track. I doubt it's Rob Northen. Probably something simpler. It may not even involve encryption.

Do you know how is the long track used? (at least, a hint?)

Oh, btw, dec0de works great for AtariST CopyLock protections, at least for Dark Side and Total Eclipse. I haven't tested with Castle Master for AtariST yet. I have documented all the steps for for the Freescape games here: https://wiki.scummvm.org/index.php?title=Freescape/DRM

**Reply by neuromancer** [reply]
I just found that [this](https://github.com/ggnkua/Atari_ST_Sources/blob/ba0237cfee64b0f5adbca6c89f255ec533b2b2de/ASM/Pompey%20Pirates/ALIEN/DEPAK_PR.O/CRAKROB4.S#L59-L63) suggests that Castle Master for AtariST is using some kind of Copylock. :thinking:

**Reply by neuromancer** [reply]
Quick update: got a message in the ScummVM FB and confirmed that dec0de works correctly on Castle Master for AtariST using `M.PRG` (however, there is another layer of compression/decryption so I need to keep digging in order to recover the original data..)
